File: 2nd/homework_2/id_69/101_SymmetricTree.py
# author: lu19kend
# https://leetcode.com/problems/symmetric-tree/description/


class Solution(object):

    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True
        queue = []
        queue.append(root.left)
        queue.append(root.right)
        while queue:
            lnode = queue.pop(0)
            rnode = queue.pop(0)
            if lnode is None and rnode is None:
                continue
            if (lnode is None) ^ (rnode is None):
                return False
            if lnode.val != rnode.val:
                return False
            queue.append(lnode.left)
            queue.append(rnode.right)
            queue.append(lnode.right)
            queue.append(rnode.left)
        return True

File: 2nd/homework_2/id_45/101_symmetric_tree.cpp
/**
 *  * Definition for a binary tree node.
 *  * struct TreeNode {
 *  *     int val;
 *  *     TreeNode *left;
 *  *     TreeNode *right;
 *  *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 *  * };
 *  */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root)
            return true;
        return compare(root->left, root->right);
    }

    bool compare(TreeNode* t1, TreeNode* t2) {
        if (!t1 && !t2)
            return true;
        if (!t1 || !t2)
            return false;
        if (t1->val != t2->val)
            return false;
        return compare(t1->left, t2->right) && compare(t1->right, t2->left);
    }
};
